src/firmware/app/control/trajectory_planner.c

```diff
@@ -421,19 +562,106 @@ void app_trajectory_planner_generateVelocityTrajectory(
+    app_trajectory_planner_generateVelocityTrajectory(&position_trajectory,
+                                                      velocity_trajectory);
+    return status;
+}
+
+
+void app_trajectory_planner_generateStatesAndReturnSegmentLengths(
+    PositionTrajectory_t* trajectory,
+    TrajectorySegment_t trajectory_segments[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS])
+{
+    // Check that the pre conditions are met
+    assert(trajectory->path_parameters.num_segments > 2);
+    assert(trajectory->path_parameters.max_allowable_linear_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_linear_speed >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_speed >= 0);
+    assert(trajectory->path_parameters.initial_linear_speed >= 0);
+    assert(trajectory->path_parameters.final_linear_speed >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_acceleration >= 0);
+    assert(trajectory->path_parameters.max_allowable_angular_speed >= 0);
+    assert(trajectory->path_parameters.num_segments <=
+           TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS);
+
+    const float t_start = trajectory->path_parameters.t_start;
+
+    PositionTrajectoryElement_t* trajectory_elements = trajectory->trajectory_elements;
+
+    trajectory_elements[0].orientation = shared_polynomial1d_getValueOrder3(
+        trajectory->path_parameters.orientation_profile, t_start);
+
+    trajectory_elements[0].position =
+        shared_polynomial2d_getValueOrder3(trajectory->path_parameters.path, t_start);
+
+    const float t_segment_size =
+        (trajectory->path_parameters.t_end - trajectory->path_parameters.t_start) /
+        (trajectory->path_parameters.num_segments - 1);
+
+    for (unsigned int i = 1; i < trajectory->path_parameters.num_segments; i++)
+    {
+        // Grab the states at each 't' value
+        trajectory_elements[i].orientation = shared_polynomial1d_getValueOrder3(
+            trajectory->path_parameters.orientation_profile,
+            t_start + i * t_segment_size);
+        trajectory_elements[i].position = shared_polynomial2d_getValueOrder3(
+            trajectory->path_parameters.path, t_start + i * t_segment_size);
+
+        // Calculate the length of each segment and store it
+        const float delta_x =
+            trajectory_elements[i].position.x - trajectory_elements[i - 1].position.x;
+        const float delta_y =
+            trajectory_elements[i].position.y - trajectory_elements[i - 1].position.y;
+
+        const float linear_segment_length = sqrt(pow(delta_x, 2) + pow(delta_y, 2));
+        const float angular_segment_length =
+            trajectory_elements[i].orientation - trajectory_elements[i - 1].orientation;
+
+        trajectory_segments[i - 1].linear_segment_length  = linear_segment_length;
+        trajectory_segments[i - 1].angular_segment_length = angular_segment_length;
+    }
+}
```

Comment:
missing newline at eof

Reply:
Added NL

Reply:
re-added the newline. Got deleted when I refactored it again.